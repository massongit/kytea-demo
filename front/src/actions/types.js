/**
 * KyTeaによる解析結果の表示Actionを表すタイプ
 * @type {string}
 */
export const SHOW_SENTENCE = "SHOW_SENTENCE"

/**
 * 品詞や読みの表示Actionを表すタイプ
 * @type {string}
 */
export const SHOW_PRONUNCIATION = "SHOW_PRONUNCIATION"

/**
 * ローディングActionを表すタイプ
 * @type {string}
 */
export const LOADING = "LOADING"
