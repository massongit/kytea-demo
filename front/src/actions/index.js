import * as types from "./types"
import {createAction} from "redux-actions"
import {intl} from "../intl"

/**
 * タグを変換する
 * @param tag タグ
 * @returns {string} タグ
 */
const convertTag = (tag) => {
    if (tag === "(Unknown)") {
        return intl.formatMessage({
            id: "unknown"
        })
    } else {
        return tag
    }
}

/**
 * 与えられたArrayに対し、map操作を行えそうなら行い、行えないようであれば空のArrayを返す
 * @param l {array}
 * @param f {function} 操作
 * @returns {array}
 */
const mapOrEmptyArray = (l, f) => {
    if (l) {
        return l.map(f)
    } else {
        return []
    }
}

/**
 * KyTeaによる解析結果の表示ActionのActionCreator
 * @type {function}
 */
export const showSentence = createAction(types.SHOW_SENTENCE, parameter => {
    // KyTeaによる解析結果の表示Action
    const action = {
        words: mapOrEmptyArray(parameter.words, w => {
            if (w) {
                return {
                    word: w.word,
                    pronunciation: mapOrEmptyArray(w.pronunciation, p => {
                        p.pronunciation = convertTag(p.pronunciation)
                        return p
                    }),
                    pos: convertTag(w.pos)
                }
            } else {
                return w
            }
        })
    }

    // 空白を除去したsentenceをActionにセット
    if (parameter.sentence) {
        action.sentence = parameter.sentence.replace(/\s/g, "")
    }

    return action
})

/**
 * 品詞や読みの表示ActionのActionCreator
 * @type {function}
 */
export const showPOSAndPronunciation = createAction(types.SHOW_PRONUNCIATION, parameter => ({
    number: parameter.number,
    word: parameter.word,
    pos: parameter.pos,
    pronunciation: parameter.pronunciation
}))
