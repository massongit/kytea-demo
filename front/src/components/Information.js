import React from "react"
import PropTypes from "prop-types"
import {Table} from "react-bootstrap"
import {FormattedNumber, intlShape} from "react-intl"

/**
 * 解析結果表示部
 */
class Information extends React.Component {
    static propTypes = {
        number: PropTypes.number.isRequired,
        word: PropTypes.string,
        pos: PropTypes.string.isRequired,
        pronunciation: PropTypes.shape({
            id: PropTypes.number,
            margin: PropTypes.number.isRequired,
            pronunciation: PropTypes.string.isRequired
        }),
        intl: intlShape.isRequired
    }

    render() {
        return (
            <Table striped bordered hover condensed>
                <thead>
                    <tr>
                        {
                            [
                                this.props.intl.formatMessage({
                                    id: "header.candidate"
                                }),
                                this.props.intl.formatMessage({
                                    id: "header.word"
                                }),
                                this.props.intl.formatMessage({
                                    id: "header.pos"
                                }),
                                this.props.intl.formatMessage({
                                    id: "header.pronunciation"
                                }),
                                this.props.intl.formatMessage({
                                    id: "header.margin"
                                })
                            ].map((header, i) => ( // 見出しを上下左右中央揃えで表示
                                <th
                                    key={i}
                                    className="text-center"
                                    style={{
                                        "verticalAlign": "middle"
                                    }}
                                >
                                    {header}
                                </th>
                            ))
                        }
                    </tr>
                </thead>
                <tbody>
                    <tr>
                        {
                            [
                                this.props.intl.formatNumber(this.props.number),
                                this.props.word,
                                this.props.pos,
                                this.props.pronunciation.pronunciation
                            ].map((data, i) => (
                                <td
                                    className="text-center"
                                    key={i}
                                >
                                    {data}
                                </td>
                            ))
                        }
                        <td className="text-right">
                            <FormattedNumber
                                value={this.props.pronunciation.margin}
                            />
                        </td>
                    </tr>
                </tbody>
            </Table>
        )
    }
}

export default Information
