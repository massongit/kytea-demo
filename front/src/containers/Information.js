import Information from "../components/Information"
import {connect} from "react-redux"
import {injectIntl} from "react-intl"

const mapStateToProps = state => ({
    number: state.showPOSAndPronunciation.number,
    word: state.showPOSAndPronunciation.word,
    pos: state.showPOSAndPronunciation.pos,
    pronunciation: state.showPOSAndPronunciation.pronunciation
})

/**
 * 解析結果表示部のContainer
 */
export default injectIntl(connect(mapStateToProps)(Information))
