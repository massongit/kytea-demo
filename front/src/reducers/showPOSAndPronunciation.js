import * as types from "../actions/types"
import {handleActions} from "redux-actions"

/**
 * 初期状態
 * @type {{word: string, pos: string, pronunciation: {pronunciation: string}}}
 */
const initialState = {
    word: "",
    pos: "",
    pronunciation: {
        pronunciation: ""
    }
}

/**
 * 品詞や読みの表示ActionのReducer
 */
export default handleActions({
    // KyTeaによる解析結果の表示Actionのとき、Stateをリセット
    [types.SHOW_SENTENCE]: () => initialState,
    // 品詞や読みの表示Actionのとき
    [types.SHOW_PRONUNCIATION]: (state, action) => {
        if (action.payload.word) { // Actionにwordが存在するとき、Stateを更新
            return {
                number: action.payload.number,
                word: action.payload.word,
                pos: action.payload.pos,
                pronunciation: action.payload.pronunciation
            }
        } else { // Actionにwordが存在しないとき、Stateを更新しない
            return state
        }
    }
}, initialState)
