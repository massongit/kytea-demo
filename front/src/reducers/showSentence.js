import * as types from "../actions/types"
import {handleActions} from "redux-actions"

/**
 * KyTeaによる解析結果の表示ActionのReducer
 */
export default handleActions(
    {
        // KyTeaによる解析結果の表示Actionのとき
        [types.SHOW_SENTENCE]: (state, action) => {
            if (action.payload.words) {
                // sentence (検査用)
                const sentence = action.payload.words.map(w => {
                    if (w && w.word) { // wordsから単語を取り出し、sentence (検査用) に追加
                        return w.word
                    } else { // words内にデータの欠損が存在する場合、Stateを更新しない
                        return undefined
                    }
                })

                // wordsから取り出した単語とsentenceが一致するとき、Stateを更新
                if (!sentence.includes(undefined) && action.payload.sentence === sentence.join("").replace(/\s/g, "")) {
                    return {
                        sentence: action.payload.sentence,
                        words: action.payload.words
                    }
                }
            }

            return state
        }
    },
    {
        sentence: "",
        words: []
    }
)
