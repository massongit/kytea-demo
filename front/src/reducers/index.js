import {combineReducers} from "redux"
import showSentence from "./showSentence"
import showPOSAndPronunciation from "./showPOSAndPronunciation"

/**
 * ルートReducer
 */
export default combineReducers({
    showSentence,
    showPOSAndPronunciation
})
