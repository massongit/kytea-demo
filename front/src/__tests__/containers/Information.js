import React from "react"
import Information from "../../containers/Information"
import deepcopy from "deepcopy"
import rootReducer from "../../reducers"
import {loadTranslation, mountWithIntl, shallowWithIntl} from "enzyme-react-intl"
import {createStore} from "redux"
import {showPOSAndPronunciation, showSentence} from "../../actions"
import {pos, pronunciation, showPOSAndPronunciationState_, showSentenceParameter, word} from "../reducers"

let store

loadTranslation("./src/translations/ja.json")

describe("containers/Information", () => {
    beforeEach(() => {
        store = createStore(rootReducer)
        store.dispatch(showSentence(deepcopy(showSentenceParameter)))
        store.dispatch(showPOSAndPronunciation(showPOSAndPronunciationState_))
    })

    it("初期状態からshowSentence, showPOSAndPronunciationとStateが遷移した際に、Componentが正しく配置されている", () => {
        const informationComponent = shallowWithIntl(
            <Information
                store={store}
            />
        ).dive()
        expect(informationComponent).toMatchSnapshot()
    })

    it("初期状態からshowSentence, showPOSAndPronunciationとStateが遷移した際に、子要素にposが含まれる", () => {
        const informationComponent = mountWithIntl(
            <Information
                store={store}
            />
        )
        expect(informationComponent.contains(pos)).toBeTruthy()
    })

    it("初期状態からshowSentence, showPOSAndPronunciationとStateが遷移した際に、子要素にwordが含まれる", () => {
        const informationComponent = mountWithIntl(
            <Information
                store={store}
            />
        )
        expect(informationComponent.contains(word)).toBeTruthy()
    })

    it("初期状態からshowSentence, showPOSAndPronunciationとStateが遷移した際に、子要素にpronunciationが含まれる", () => {
        const informationComponent = mountWithIntl(
            <Information
                store={store}
            />
        )
        for (const w of pronunciation) {
            expect(informationComponent.contains(w.pronunciation)).toBeTruthy()
        }
    })
})
