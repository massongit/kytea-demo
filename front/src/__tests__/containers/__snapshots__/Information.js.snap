// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Information/snapshot 初期状態からshowSentence, showPOSAndPronunciationとStateが遷移した際に、Componentが正しく配置されている 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Information)
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "abstract.kyTea": "京都テキスト解析ツールキット",
          "candidate": "候補{no}",
          "description.input": "日本語の文章を入力し、予測ボタンをクリックしてください。",
          "description.kyTea": "日本語の文章中に存在する各単語の品詞や読みを予測するシステムです。",
          "description.result": "単語を選択すると、その単語の品詞や読みが表示されます。",
          "errorMessage.default": "エラーが発生しました: {message}",
          "errorMessage.predict": "予測中にエラーが発生しました: {message}",
          "header.candidate": "候補",
          "header.margin": "SVM平面からのマージン (読み)",
          "header.pos": "品詞",
          "header.pronunciation": "読み",
          "header.word": "単語",
          "predict": "予測",
          "title.description": "概要",
          "title.input": "入力",
          "title.result": "結果",
          "unknown": "(予測不能)",
        },
        "now": [Function],
        "onError": [Function],
        "textComponent": "span",
        "timeZone": null,
      }
    }
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "dispatch": [Function],
      "intl": Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "abstract.kyTea": "京都テキスト解析ツールキット",
          "candidate": "候補{no}",
          "description.input": "日本語の文章を入力し、予測ボタンをクリックしてください。",
          "description.kyTea": "日本語の文章中に存在する各単語の品詞や読みを予測するシステムです。",
          "description.result": "単語を選択すると、その単語の品詞や読みが表示されます。",
          "errorMessage.default": "エラーが発生しました: {message}",
          "errorMessage.predict": "予測中にエラーが発生しました: {message}",
          "header.candidate": "候補",
          "header.margin": "SVM平面からのマージン (読み)",
          "header.pos": "品詞",
          "header.pronunciation": "読み",
          "header.word": "単語",
          "predict": "予測",
          "title.description": "概要",
          "title.input": "入力",
          "title.result": "結果",
          "unknown": "(予測不能)",
        },
        "now": [Function],
        "onError": [Function],
        "textComponent": "span",
        "timeZone": null,
      },
      "number": 1,
      "pos": "名詞",
      "pronunciation": Object {
        "id": 1,
        "margin": 100,
        "pronunciation": "やきゅう",
      },
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      },
      "word": "野球",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "intl": Object {
          "defaultFormats": Object {},
          "defaultLocale": "en",
          "formatDate": [Function],
          "formatHTMLMessage": [Function],
          "formatMessage": [Function],
          "formatNumber": [Function],
          "formatPlural": [Function],
          "formatRelative": [Function],
          "formatTime": [Function],
          "formats": Object {},
          "formatters": Object {
            "getDateTimeFormat": [Function],
            "getMessageFormat": [Function],
            "getNumberFormat": [Function],
            "getPluralFormat": [Function],
            "getRelativeFormat": [Function],
          },
          "locale": "en",
          "messages": Object {
            "abstract.kyTea": "京都テキスト解析ツールキット",
            "candidate": "候補{no}",
            "description.input": "日本語の文章を入力し、予測ボタンをクリックしてください。",
            "description.kyTea": "日本語の文章中に存在する各単語の品詞や読みを予測するシステムです。",
            "description.result": "単語を選択すると、その単語の品詞や読みが表示されます。",
            "errorMessage.default": "エラーが発生しました: {message}",
            "errorMessage.predict": "予測中にエラーが発生しました: {message}",
            "header.candidate": "候補",
            "header.margin": "SVM平面からのマージン (読み)",
            "header.pos": "品詞",
            "header.pronunciation": "読み",
            "header.word": "単語",
            "predict": "予測",
            "title.description": "概要",
            "title.input": "入力",
            "title.result": "結果",
            "unknown": "(予測不能)",
          },
          "now": [Function],
          "onError": [Function],
          "textComponent": "span",
          "timeZone": null,
        },
        "number": 1,
        "pos": "名詞",
        "pronunciation": Object {
          "id": 1,
          "margin": 100,
          "pronunciation": "やきゅう",
        },
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
          "unsubscribe": [Function],
        },
        "word": "野球",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "intl": Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "abstract.kyTea": "京都テキスト解析ツールキット",
          "candidate": "候補{no}",
          "description.input": "日本語の文章を入力し、予測ボタンをクリックしてください。",
          "description.kyTea": "日本語の文章中に存在する各単語の品詞や読みを予測するシステムです。",
          "description.result": "単語を選択すると、その単語の品詞や読みが表示されます。",
          "errorMessage.default": "エラーが発生しました: {message}",
          "errorMessage.predict": "予測中にエラーが発生しました: {message}",
          "header.candidate": "候補",
          "header.margin": "SVM平面からのマージン (読み)",
          "header.pos": "品詞",
          "header.pronunciation": "読み",
          "header.word": "単語",
          "predict": "予測",
          "title.description": "概要",
          "title.input": "入力",
          "title.result": "結果",
          "unknown": "(予測不能)",
        },
        "now": [Function],
        "onError": [Function],
        "textComponent": "span",
        "timeZone": null,
      },
    },
  },
}
`;
